packages/library-base/src/commonMain/kotlin/io/realm/query/RealmQuery.kt

```diff
@@ -0,0 +1,393 @@
+ * objects in the order they where inserted. Use the [sort] functions if a specific order is
+ * required.
+ *
+ * Results are obtained quickly most of the times when using [find]. However, launching heavy
+ * queries from the UI thread may result in a drop of frames or even ANRs. If you want to prevent
+ * these behaviors, you can use [asFlow] and collect the results asynchronously.
+ *
+ * @param T the class of the objects to be queried.
+ */
+interface RealmQuery<T : RealmObject> : RealmElementQuery<T> {
+
+    /**
+     * Appends the query represented by [filter] to an existing query using a logical `AND`.
+     *
+     * @param filter the Realm Query Language predicate to append.
+     * @param arguments Realm values for the predicate.
+     */
+    fun query(filter: String, vararg arguments: Any?): RealmQuery<T>
+
+    /**
+     * Sorts the query result by the specific property name according to [sortOrder], which is
+     * [Sort.ASCENDING] by default.
+     *
+     * Sorting is currently limited to character sets in 'Latin Basic', 'Latin Supplement',
+     * 'Latin Extended A', and 'Latin Extended B' (UTF-8 range 0-591). For other character sets
+     * sorting will have no effect.
+     *
+     * @param property the property name to sort by.
+     * @throws [IllegalArgumentException] if the property name does not exist.
+     */
+    fun sort(property: String, sortOrder: Sort = Sort.ASCENDING): RealmQuery<T>
+
+    /**
+     * Sorts the query result by the specific property name according to [Pair]s of properties and
+     * sorting order.
+     *
+     * Sorting is currently limited to character sets in 'Latin Basic', 'Latin Supplement',
+     * 'Latin Extended A', and 'Latin Extended B' (UTF-8 range 0-591). For other character sets
+     * sorting will have no effect.
+     *
+     * @param propertyAndSortOrder pair containing the property and the order to sort by.
+     * @throws [IllegalArgumentException] if the property name does not exist.
+     */
+    fun sort(
+        propertyAndSortOrder: Pair<String, Sort>,
+        vararg additionalPropertiesAndOrders: Pair<String, Sort>
+    ): RealmQuery<T>
+
+    /**
+     * Selects a distinct set of objects of a specific class. When multiple distinct fields are
+     * given, all unique combinations of values in the fields will be returned. In case of multiple
+     * matches, it is undefined which object is returned. Unless the result is sorted, then the
+     * first object will be returned.
+     *
+     * @param property first field name to use when finding distinct objects.
+     * @param extraProperties remaining property names when determining all unique combinations of
+     * field values.
+     * @throws IllegalArgumentException if [property] does not exist, is an unsupported type, or
+     * points to a linked field.
+     */
+    // TODO "points to a linked field" doesn't apply yet
```

Comment:
I am not completely certain about what is missing to get this to work, so a bit hard to create the issue. But, I think this will be revisited as part of fixing the actual use case and writing tests for it, so the risk of missing this is not that high. Will just leave it as is. Maybe @edualonso can clarify when he is back.